crates/oxc_transformer/src/es2017/async_to_generator.rs

```diff
@@ -208,24 +397,306 @@ impl<'a, 'ctx> AsyncToGenerator<'a, 'ctx> {
             *statement = ctx.ast.statement_return(expression.span(), Some(expression));
         }
 
-        let r#type = FunctionType::FunctionExpression;
-        let parameters = ctx.ast.alloc(ctx.ast.formal_parameters(
+        let params = ctx.alloc(ctx.ast.move_formal_parameters(&mut arrow.params));
+        let generator_function_id = arrow.scope_id.get().unwrap();
+        ctx.scopes_mut().get_flags_mut(generator_function_id).remove(ScopeFlags::Arrow);
+
+        if !Self::is_function_length_affected(&params) {
+            return Some(self.create_async_to_generator_call(
+                params,
+                ctx.ast.alloc(body),
+                generator_function_id,
+                ctx,
+            ));
+        }
+
+        let wrapper_scope_id = ctx.create_child_scope(ctx.current_scope_id(), ScopeFlags::Function);
+        // The generator function will move to inside wrapper, so we need
+        // to change the parent scope of the generator function to the wrapper function.
+        ctx.scopes_mut().change_parent_id(generator_function_id, Some(wrapper_scope_id));
+
+        let bound_ident = Self::create_bound_identifier(None, wrapper_scope_id, ctx);
+
+        let caller_function = {
+            let scope_id = ctx.create_child_scope(wrapper_scope_id, ScopeFlags::Function);
+            let params = Self::create_placeholder_params(&params, scope_id, ctx);
+            let statements = ctx.ast.vec1(Self::create_apply_call_statement(&bound_ident, ctx));
+            let body = ctx.ast.alloc_function_body(SPAN, ctx.ast.vec(), statements);
+            let id = Self::infer_function_id_from_variable_declarator(wrapper_scope_id, ctx);
+            let function = Self::create_function(id, params, body, scope_id, ctx);
+            let argument = Some(ctx.ast.expression_from_function(function));
+            ctx.ast.statement_return(SPAN, argument)
+        };
+
+        // Wrapper function
+        {
+            let statement = self.create_async_to_generator_declaration(
+                &bound_ident,
+                params,
+                ctx.ast.alloc(body),
+                generator_function_id,
+                ctx,
+            );
+            let mut statements = ctx.ast.vec_with_capacity(2);
+            statements.push(statement);
+            statements.push(caller_function);
+            let body = ctx.ast.alloc_function_body(SPAN, ctx.ast.vec(), statements);
+            let params = Self::create_empty_params(ctx);
+            let wrapper_function = Self::create_function(None, params, body, wrapper_scope_id, ctx);
+            // Construct the IIFE
+            let callee = ctx.ast.expression_from_function(wrapper_function);
+            Some(ctx.ast.expression_call(SPAN, callee, NONE, ctx.ast.vec(), false))
+        }
+    }
+
+    /// Infers the function id from [`Ancestor::VariableDeclaratorInit`].
+    #[inline]
```

Comment:
What's the reason for `#[inline]` here?

Generally it's better to leave decisions on whether to inline or not to the compiler, except when:

1. the function is trivial (only a couple of assembly instructions) or
2. there's some specific circumstances where you know more than the compiler, or want to "tell" the compiler something.

An example of (2) is when the function needs to be inlined for the compiler to prove an `unwrap()` will never panic and that it can remove it:

```rs
fn caller(expr: &Expression) {
    if matches!(expr, Expression::BinaryExpression(_) {
        // `expr` is definitely a `BinaryExpression`
        callee(expr);
    }
}

// `#[inline]` so that compiler can prove that `expr` is always a `BinaryExpression`
// and remove the `unreachable!()` branch
#[inline]
fn callee(expr: &Expression) {
    // If this function isn't inlined, the compiler knows nothing about the type of `expr`
    // because each function is optimized separately.
    // But if this function is inlined into `caller`, then compiler sees all the code together.
    // It knows `expr` is a `BinaryExpression` because of the `if` statement in `caller`,
    // so it will remove the `else { unreachable!() }` branch here.
    let Expression::BinaryExpression(bin_expr) = expr else { unreachable!() };
}
```

If you're doing something like this, include a comment to explain why `#[inline]` (like the comment above `callee` in example above).

But my rule of thumb is that if I'm not sure whether to inline or not, don't. I trust the compiler more than myself!

Same for most of the other functions below.

Reply:
Thanks for explaining a lot, I understand now!